geode-core/src/main/java/org/apache/geode/cache/query/internal/QueryConfigurationServiceImpl.java

```diff
@@ -0,0 +1,176 @@
+      this.authorizer = new RestrictedMethodAuthorizer(cache);
+    }
+    return true;
+  }
+
+  @Override
+  public Class<? extends CacheService> getInterface() {
+    return QueryConfigurationService.class;
+  }
+
+  @Override
+  public CacheServiceMBeanBase getMBean() {
+    return null;
+  }
+
+  @Override
+  public void close() {
+
+  }
+
+  @Override
+  public MethodInvocationAuthorizer getMethodAuthorizer() {
+    return authorizer;
+  }
+
+  @Override
+  public void updateMethodAuthorizer(Cache cache, QueryMethodAuthorizerCreation creation) {
+    updateMethodAuthorizer(cache, creation.getClassName(), creation.getParameters());
+  }
+
+  @Override
+  public void updateMethodAuthorizer(Cache cache, String className, Set<String> parameters) {
+    if (isSecurityDisabled((InternalCache) cache) || ALLOW_UNTRUSTED_METHOD_INVOCATION) {
+      return;
+    }
+
+    if (className == null) {
+      logError(UPDATE_ERROR_MESSAGE + NULL_CLASS_NAME + AUTHORIZER_NOT_UPDATED);
+      return;
+    }
+
+    if (className.equals(RestrictedMethodAuthorizer.class.getSimpleName())) {
+      this.authorizer = new RestrictedMethodAuthorizer(cache);
+    } else if (className.equals(UnrestrictedMethodAuthorizer.class.getSimpleName())) {
+      this.authorizer = new UnrestrictedMethodAuthorizer(cache);
+    } else if (className.equals(JavaBeanAccessorMethodAuthorizer.class.getSimpleName())) {
+      this.authorizer = new JavaBeanAccessorMethodAuthorizer(cache, parameters);
+    } else if (className.equals(RegExMethodAuthorizer.class.getSimpleName())) {
+      this.authorizer = new RegExMethodAuthorizer(cache, parameters);
+    } else {
+      try {
+        @SuppressWarnings("unchecked")
+        Class<MethodInvocationAuthorizer> userClass =
+            (Class<MethodInvocationAuthorizer>) ClassPathLoader.getLatest().forName(className);
+        Constructor<MethodInvocationAuthorizer> userConstructor =
+            userClass.getDeclaredConstructor(Cache.class, Set.class);
+        this.authorizer = userConstructor.newInstance(cache, parameters);
+      } catch (Exception e) {
+        logErrorMessage(className, e);
+      }
+    }
+  }
+
+  private boolean isSecurityDisabled(InternalCache cache) {
+    return !cache.getSecurityService().isIntegratedSecurity();
+  }
+
+  private void logErrorMessage(String className, Exception e) {
+    if (e instanceof ClassNotFoundException) {
+      logError(UPDATE_ERROR_MESSAGE + NO_CLASS_FOUND + className + ". " + AUTHORIZER_NOT_UPDATED
+          + " Reason: " + e.toString());
+    } else if (e instanceof NoSuchMethodException || e instanceof SecurityException) {
+      logError(UPDATE_ERROR_MESSAGE + NO_VALID_CONSTRUCTOR + AUTHORIZER_NOT_UPDATED + " Reason: "
+          + e.toString());
+    } else {
+      logError(UPDATE_ERROR_MESSAGE + INSTANTIATION_ERROR + AUTHORIZER_NOT_UPDATED + " Reason: "
+          + e.toString());
+    }
+  }
+
+  // For testing
+  void logError(String message) {
+    logger.error(message);
+  }
```

Comment:
By the time the exception is thrown, the XML has already been entirely parsed without any problems and the cache is in the process of being configured. Even if an exception is thrown here, the `QueryConfigurationServiceImpl` will still be available and have the default `RestrictedMethodAuthorizer` configured for use. Perhaps this is grounds for aborting startup, I don't know how other cache elements handle similar issues.

+1 to printing out the stack trace though. Making errors more visible seems like it's rarely a bad thing.

Reply:
I would be hesitant to abort the startup of the cluster entirely because this can put people in situations where their whole cluster is held hostage by a feature they're not particularly invested in. Since our default settings are to restrict access to methods I would think that a cluster starting with the wrong authorizer due to a bug or user error could be fixed by using our gfsh command (or other methods in the future) to set the authorizer to the one they intended is preferable to a cluster failing to start. In the past this is how I've seen errors like this handled in cases such as the datasources loaded from cluster config which were used by the JDBC connector, and although this is a security related setting we default to high security rather than low security so I don't see that being an issue.

Reply:
@BenjaminPerryRoss: I don't agree on this one, if I configure something wrong and the `cache.xml` can't be parsed altogether, I expect the member startup to halt right away and throw an exception so I know I did something wrong. If we continue with the startup and use the default restricted authorizer, it will be harder for our users later in time to debug what's wrong (_hey, why is this method being rejected if the authorizer I've provided is supposed to allow it??_ ). We also need to consider that the user executing queries (developer) might not the same one as the one configuring the cluster in the first place (operator).
That said, throwing an exception and stopping the member from starting up is exactly what we do when a user makes a mistake configuring other aspects of the cache, like a `CacheListener`, a `CacheLoader`, a `CacheWriter`, etc..

Reply:
Sorry, missed this comment. While Region Attributes like CacheListener, CacheWriter, etc do cause a failure if they're misconfigured or some other problem occurs during configuration there are also features which only log errors like JNDIBindings. My understanding was that whether or not a configuration halts the start up of a member or not is related to the importance of the object being configured. I don't feel too strongly about this specific object although my original inclination was definitely that it wasn't critical enough to prevent startup so if you feel like it's important to change that then we can definitely do that. I mostly just want to make sure we're weighing the value of the Method Authorizer coming up with configuration vs the cost of failing to start a member in these situations.

Reply:
I also wanted to clarify - if the cache.xml has invalid syntax the parsing will fail before it reaches our custom code. The only circumstance in which we log a warning is if the xml is defined in such a way that the xml is parsed correctly, but an exception is thrown while trying to build the QueryConfigurationService/MethodAuthorizer from the values parsed.

Reply:
Hey @BenjaminPerryRoss , thanks for answering my concerns.
I'd still argue that it's quite important to totally halt the startup if we can't instantiate the class the customer has configured and expects to use. If we just log a warning message and continue as it nothing bad had happened, it'll be much harder for them later in time to discover what went wrong after a query failure due to an non authorised method... it could get even worse if they "lose" the initial section of the log due to rolling (I've seen it happening).
Maybe @nabarunnag or @jhuynh1 have some thoughts on this?.

Reply:
I think Juan's concerns have swayed my opinion to agree with making the user aware that a class is missing or configured improperly.  So at this point I would agree with Juan's concern.

Reply:
This change has been applied in commit 6df1e47. Cluster startup is now aborted if an exception is thrown during method authorizer initialization.